README.md
# embedded-scd [![CircleCI](https://circleci.com/gh/Sensirion/embedded-scd.svg?style=shield)](https://circleci.com/gh/Sensirion/embedded-scd) [![GitHub license](https://img.shields.io/badge/license-BSD3-blue.svg)](https://raw.githubusercontent.com/Sensirion/embedded-scd/master/LICENSE)
This repository contains the embedded driver sources for Sensirion's SCD product
line.

## Clone this repository
```
 git clone --recursive https://github.com/Sensirion/embedded-scd.git
```

## Repository content
* embedded-common (submodule repository for the common embedded driver HAL)
* scd30 (SCD30 driver related)

## Collecting resources
```
make release
```
This will create the `release` folder with the necessary driver files in it,
including a Makefile. That way, you have just one folder with all the sources
ready to build your driver for your platform.

## Files to adjust (from embedded-common)
You only need to touch the following files:

* `sensirion_arch_config.h` (architecture specifics, you need to specify
the integer sizes)

and depending on your i2c implementation either of the following:

* `embedded-common/hw_i2c/sensirion_hw_i2c_implementation.c`
  functions for hardware i2c communication if your platform supports that
* `embedded-common/sw_i2c/sensirion_sw_i2c_implementation.c`
  functions for software i2c communication via GPIOs

## Building the driver
1. Adjust sensirion\_arch\_config.h if you don't have the `<stdint.h>` header
   file available
2. Implement necessary functions in one of the `*_implementation.c` files
   described above
3. make

---

Please check the [embedded-common](https://github.com/Sensirion/embedded-common)
repository for further information and sample implementations.

---


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.1.2] - 2020-12-14

 * [`changed`] Updated embedded-common to 0.1.0 to improve compatibility when
               using multiple embedded drivers

## [2.1.1] - 2020-11-05

 * [`changed`] Only include needed files from embedded-common in release zip
 * [`fixed`]   Fix bug where data ready flag could not be read.
 * [`fixed`]   Fix readout error with drifting clocks by polling data ready
               flag.

## [2.1.0] - 2020-07-08

 * [`fixed`]    Fix typo in header include-guard.
 * [`changed`]  Use `float` instead of the custom `float32_t` type
 * [`changed`]  Use configuration independent endianness conversions: no need to
                define SENSIRION_BIG_ENDIAN anymore.

## [2.0.2] - 2019-11-01

 * [`fixed`]    SCD30: Fix buffer overflow when reading measurements that was
                introduced in 2.0.1.

## [2.0.1] - 2019-10-17

 * [`fixed`]    Copy correct `AUTHORS`, `CHANGELOG.md`, `LICENSE`, and
                `README.md` files to target locations when running the `release`
                target of the driver's root Makefile.
 * [`removed`]  Remove the `AUTHORS` file from the driver and the
                `embedded-common` submodule, as it adds more noise than benefit.
                The contributors can be found in the git log.
 * [`changed`]  Avoid converting twice between endianness in
                `scd30_read_measurement`. (improves performance and portablity)
 * [`changed`]  Cleanup scd30 example

## [2.0.0] - 2019-10-01

 * [`changed`]  Functions are now dedicated per sensor (e.g. `scd30_probe()`
                instead of `scd_probe()`)
 * [`changed`]  Move the i2c init call out of `probe()` and into the example
 * [`added`]    Add `scd30_read_serial()` to read out the serial number
 * [`added`]    embedded-common: Add multi-bus support along with an example.
 * [`fixed`]    Fix compilation warnings when compiling the linux user space
                sample implementation with `CONFIG_I2C_TYPE` set to `sw_i2c`
 * [`changed`]  Split out `default_config.inc` from Makefile to configure paths
                and CFLAGS for the SCD30 driver
 * [`changed`]  Only one example with either `hw_i2c` or `sw_i2c` is build,
                depending on `CONFIG_I2C_TYPE`. Defaults to `hw_i2c`.
 * [`fixed`]    Fix clock stretching timeout that might occur when
                `CONFIG_I2C_TYPE` is set to `sw_i2c`

## [1.0.0] - 2019-05-14

 * `[fixed]` Add processing delay on i2c writes
 * `[fixed]` Potential issues regarding strict aliasing
 * `[changed]` Use stdint types (`uint16_t` instead of `u16`)

## [0.9.0] - 2018-10-08

 * `[fixed]` Measurement readout
 * `[added]` Self calibration
   * `scd_get_automatic_self_calibration`
   * `scd_set_forced_recalibration`
   * `scd_enable_automatic_self_calibration`
 * `[fixed]` Parameters on big-endian systems

## [0.1.0] - 2018-07-20

 * Initial support for SCD30

[Unreleased]: https://github.com/Sensirion/embedded-scd/compare/2.1.2...master
[2.1.2]: https://github.com/Sensirion/embedded-scd/compare/2.1.1...2.1.2
[2.1.1]: https://github.com/Sensirion/embedded-scd/compare/2.1.0...2.1.1
[2.1.0]: https://github.com/Sensirion/embedded-scd/compare/2.0.2...2.1.0
[2.0.2]: https://github.com/Sensirion/embedded-scd/compare/2.0.1...2.0.2
[2.0.1]: https://github.com/Sensirion/embedded-scd/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/Sensirion/embedded-scd/compare/1.0.0...2.0.0
[1.0.0]: https://github.com/Sensirion/embedded-scd/compare/0.9.0...1.0.0
[0.9.0]: https://github.com/Sensirion/embedded-scd/compare/0.1.0...0.9.0
[0.1.0]: https://github.com/Sensirion/embedded-scd/releases/tag/0.1.0


tests/scd-test.cpp
#include "scd30.h"
#include "sensirion_test_setup.h"

#define SCD30_TEST_MIN_PRESSURE 700
#define SCD30_TEST_MAX_PRESSURE 1400
#define SCD30_TEST_MIN_MEASUREMENT_INTERVAL 2
#define SCD30_TEST_MAX_MEASUREMENT_INTERVAL 1800
#define SCD30_MIN_CO2 0
#define SCD30_MAX_CO2 40000
#define SCD30_MIN_TEMPERATURE 10
#define SCD30_MAX_TEMPERATURE 35
#define SCD30_MIN_HUMIDITY 20
#define SCD30_MAX_HUMIDITY 100

static void scd30_tests(uint16_t interval_s, uint16_t pressure_mbar) {
    int16_t ret;
    uint16_t data_ready, timeout;
    float co2_ppm;
    float temperature;
    float humidity;

    printf("\t\t\t\tTesting with pressure = %umbar\n", pressure_mbar);

    ret = scd30_start_periodic_measurement(pressure_mbar);
    if (pressure_mbar == 0 || (pressure_mbar >= SCD30_TEST_MIN_PRESSURE &&
                               pressure_mbar <= SCD30_TEST_MAX_PRESSURE)) {
        CHECK_ZERO_TEXT(ret,
                        "scd30_start_periodic_measurement should not fail");
    } else {
        CHECK_TRUE_TEXT(ret, "scd30_start_periodic_measurement should fail");
        return;
    }

    // Check if there is old data available
    ret = scd30_get_data_ready(&data_ready);
    CHECK_ZERO_TEXT(ret, "scd30_get_data_ready for old data");

    if (data_ready) {
        // Get old data and discard it
        ret = scd30_read_measurement(&co2_ppm, &temperature, &humidity);
        CHECK_ZERO_TEXT(ret, "scd30_read_measurement for old data");
    } else {
        // Ensure that reads outside the readiness window fail
        ret = scd30_read_measurement(&co2_ppm, &temperature, &humidity);
        CHECK_EQUAL_TEXT(SCD30_STATUS_NOT_READY, ret,
                         "scd30_read_measurement not ready");
    }

    // Sleep for interval_s seconds minus 100ms
    sensirion_sleep_usec((1e6 * interval_s) - 1e5);

    // Wait for data to be ready
    for (timeout = 10; timeout > 0; --timeout) {
        sensirion_sleep_usec(50000);  // Sleep 50ms
        ret = scd30_get_data_ready(&data_ready);
        CHECK_ZERO_TEXT(ret, "scd30_get_data_ready while polling");
        if (data_ready)
            break;
    }
    CHECK_TRUE_TEXT(timeout, "scd30_get_data_ready not ready");

    ret = scd30_read_measurement(&co2_ppm, &temperature, &humidity);
    CHECK_ZERO_TEXT(ret, "scd30_read_measurement");
    printf("\t\t\t\t\tCO2 = %fppm\n", co2_ppm);
    printf("\t\t\t\t\tTmp = %f degree C\n", temperature);
    printf("\t\t\t\t\t%%RH = %f%%\n", humidity);
    CHECK_TRUE_TEXT(co2_ppm >= SCD30_MIN_CO2 && co2_ppm <= SCD30_MAX_CO2,
                    "scd30_measurement CO2");
    CHECK_TRUE_TEXT(temperature >= SCD30_MIN_TEMPERATURE &&
                        temperature <= SCD30_MAX_TEMPERATURE,
                    "scd30_measurement temperature");
    CHECK_TRUE_TEXT(humidity >= SCD30_MIN_HUMIDITY &&
                        humidity <= SCD30_MAX_HUMIDITY,
                    "scd30_measurement humidity");

    ret = scd30_stop_periodic_measurement();
    CHECK_ZERO_TEXT(ret, "scd30_stop_periodic_measurement");
}

static void scd30_test_all_pressures(uint16_t interval_s) {
    scd30_tests(interval_s, 0);
    scd30_tests(interval_s, SCD30_TEST_MIN_PRESSURE - 1);
    scd30_tests(interval_s, SCD30_TEST_MIN_PRESSURE);
    scd30_tests(interval_s, SCD30_TEST_MAX_PRESSURE);
    scd30_tests(interval_s, SCD30_TEST_MAX_PRESSURE + 1);
}

static void scd30_test_set_measurement_interval(uint16_t interval_s) {
    printf("\t\t\tTesting measurement interval = %us\n", interval_s);
    int16_t ret = scd30_set_measurement_interval(interval_s);
    if (interval_s >= SCD30_TEST_MIN_MEASUREMENT_INTERVAL &&
        interval_s <= SCD30_TEST_MAX_MEASUREMENT_INTERVAL) {
        CHECK_ZERO_TEXT(ret, "scd30_set_measurement_interval should work");
        scd30_test_all_pressures(interval_s);
    } else {
        CHECK_TRUE_TEXT(ret, "scd30_set_measurement_interval should fail");
    }
}

static void scd30_test_all_measurement_intervals() {
    scd30_test_set_measurement_interval(SCD30_TEST_MIN_MEASUREMENT_INTERVAL -
                                        1);
    scd30_test_set_measurement_interval(SCD30_TEST_MIN_MEASUREMENT_INTERVAL);
    scd30_test_set_measurement_interval(2 *
                                        SCD30_TEST_MIN_MEASUREMENT_INTERVAL);
    scd30_test_set_measurement_interval(SCD30_TEST_MAX_MEASUREMENT_INTERVAL +
                                        1);
}

static void scd30_test_temperature_offset(uint16_t temp_offset) {
    printf("\t\tTesting temperature offset = %u degree C\n", temp_offset);
    int16_t ret = scd30_set_temperature_offset(temp_offset * 100);
    CHECK_ZERO_TEXT(ret, "scd30_set_temperature_offset");
    scd30_test_all_measurement_intervals();
}

static void scd30_test_all_temperature_offsets() {
    scd30_test_temperature_offset(0);   // No offset
    scd30_test_temperature_offset(50);  // 50 degrees C
}

static void scd30_test_altitude(uint16_t altitude) {
    int16_t ret;
    printf("\tTesting altitude = %um\n", altitude);
    ret = scd30_set_altitude(altitude);
    CHECK_ZERO_TEXT(ret, "scd30_set_altitude");
    scd30_test_all_temperature_offsets();
}

static void scd30_test_all_altitudes() {
    scd30_test_altitude(0);    // Altitude compensation disabled
    scd30_test_altitude(500);  // With altitude compensation on 500m
}

static void scd30_test_automatic_calibration(uint8_t set_asc) {
    uint8_t get_asc;
    int16_t ret;

    ret = scd30_enable_automatic_self_calibration(set_asc);
    CHECK_ZERO_TEXT(ret, "scd30_enable_automatic_self_calibration");

    ret = scd30_get_automatic_self_calibration(&get_asc);
    CHECK_ZERO_TEXT(ret, "scd30_get_automatic_self_calibration");

    CHECK_EQUAL_TEXT(set_asc, get_asc,
                     "automatic_self_calibration variables should match");
    scd30_test_all_altitudes();
}

static int16_t scd30_reset() {
    const uint16_t arg = 1;
    int16_t ret = sensirion_i2c_write_cmd_with_args(0x61, 0xD304, &arg,
                                                    SENSIRION_NUM_WORDS(arg));
    sensirion_sleep_usec(1e6);
    return ret;
}

static void scd30_test_setup() {
    int16_t ret;
    char serial[33];

    ret = scd30_reset();
    CHECK_ZERO_TEXT(ret, "scd30_reset");

    const char* version = scd30_get_driver_version();
    printf("scd30_get_driver_version: %s\n", version);

    uint8_t addr = scd30_get_configured_address();
    CHECK_EQUAL_TEXT(0x61, addr, "scd30_get_configured_address");

    ret = scd30_read_serial(serial);
    CHECK_ZERO_TEXT(ret, "scd30_get_serial_id");
    printf("SCD30 serial: %s\n", serial);
}

TEST_GROUP (SCDTestGroup) {
    void setup() {
        int16_t ret;
        sensirion_i2c_init();
        ret = sensirion_i2c_mux_set_single_channel(0x72, 3);
        CHECK_EQUAL_TEXT(0, ret, "sensirion_i2c_mux_select_TEXT(0x72, 3)");
        ret = scd30_probe();
        CHECK_ZERO_TEXT(ret, "scd30_probe in test setup");
        scd30_test_setup();
    }

    void teardown() {
        int16_t ret = scd30_reset();
        CHECK_EQUAL_TEXT(0, ret, "scd30_reset");
        sensirion_i2c_release();
    }
};

TEST (SCDTestGroup, SCD30Test_no_automatic) {
    printf("Starting SCD30 Test without automatic calibration\n");
    scd30_test_automatic_calibration(0);
}

TEST (SCDTestGroup, SCD30Test_with_automatic) {
    printf("Starting SCD30 Test with automatic calibration\n");
    scd30_test_automatic_calibration(1);
}

TEST (SCDTestGroup, SCD30Test_with_forced) {
    printf("Starting SCD30 Test with forced calibration\n");
    int16_t ret = scd30_set_forced_recalibration(400);
    CHECK_ZERO_TEXT(ret, "scd30_set_forced_recalibration");
    scd30_test_all_altitudes();
}


scd30/scd30.h
/*
 * Copyright (c) 2018, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SCD30_H
#define SCD30_H
#include "sensirion_arch_config.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"

#define SCD30_STATUS_NOT_READY 2

#ifdef __cplusplus
extern "C" {
#endif

/**
 * scd30_probe() - check if the SCD sensor is available and initialize it
 *
 * @return  0 on success, an error code otherwise.
 */
int16_t scd30_probe(void);

/**
 * scd30_get_driver_version() - Returns the driver version
 *
 * @return  Driver version string
 */
const char* scd30_get_driver_version(void);

/**
 * scd30_get_configured_address() - Returns the configured I2C address
 *
 * @return      uint8_t I2C address
 */
uint8_t scd30_get_configured_address(void);

/**
 * scd30_start_periodic_measurement() - Start continuous measurement to measure
 * CO2 concentration, relative humidity and temperature or updates the ambient
 * pressure if the periodic measurement is already running.
 *
 * Measurement data which is not read from the sensor is continuously
 * overwritten. The CO2 measurement value can be compensated for ambient
 * pressure by setting the pressure value in mBar. Setting the ambient pressure
 * overwrites previous and future settings of altitude compensation. Setting the
 * pressure to zero deactivates the ambient pressure compensation.
 * The continuous measurement status is saved in non-volatile memory. The last
 * measurement mode is resumed after repowering.
 *
 * @param ambient_pressure_mbar Ambient pressure in millibars. 0 to deactivate
 *                              ambient pressure compensation (reverts to
 *                              altitude compensation, if set), 700-1200mBar
 *                              allowable range otherwise
 *
 * @return                      0 if the command was successful, an error code
 *                              otherwise
 */
int16_t scd30_start_periodic_measurement(uint16_t ambient_pressure_mbar);

/**
 * scd30_stop_periodic_measurement() - Stop the continuous measurement
 *
 * @return  0 if the command was successful, else an error code
 */
int16_t scd30_stop_periodic_measurement(void);

/**
 * scd30_read_measurement() - Read out an available measurement when new
 * measurement data is available.
 * Make sure that the measurement is completed by reading the data ready status
 * bit with scd30_get_data_ready().
 *
 * @param co2_ppm       CO2 concentration in ppm
 * @param temperature   the address for the result of the temperature
 *                      measurement
 * @param humidity      the address for the result of the relative humidity
 *                      measurement
 *
 * @return              0 if the command was successful, an error code otherwise
 *                      The error code SCD30_STATUS_NOT_READY is returned if the
 *                      measurement is retrieved outside of the 100ms readiness
 *                      window after the sensor's internal measurement. With
 *                      imprecise or drifting clocks, this could occasionally
 *                      occur.
 */
int16_t scd30_read_measurement(float* co2_ppm, float* temperature,
                               float* humidity);

/**
 * scd30_set_measurement_interval() - Sets the measurement interval in
 * continuous measurement mode.
 *
 * The initial value on powerup is 2s. The chosen measurement interval is saved
 * in non-volatile memory and thus is not reset to its initial value after power
 * up.
 *
 * @param interval_sec  The measurement interval in seconds. The allowable range
 *                      is 2-1800s
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd30_set_measurement_interval(uint16_t interval_sec);

/**
 * scd30_get_data_ready() - Get data ready status
 *
 * Data ready command is used to determine if a measurement can be read from the
 * sensor's buffer. Whenever there is a measurement available from the internal
 * buffer this command returns 1 and 0 otherwise. As soon as the measurement has
 * been read by the return value changes to 0. It is recommended to use the data
 * ready status byte before readout of the measurement values with
 * scd30_read_measurement().
 *
 * @param data_ready    Pointer to memory of where to set the data ready bit.
 *                      The memory is set to 1 if a measurement is ready to be
 *                      fetched, 0 otherwise.
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd30_get_data_ready(uint16_t* data_ready);

/**
 * scd30_set_temperature_offset() - Set the temperature offset
 *
 * The on-board RH/T sensor is influenced by thermal self-heating of SCD30 and
 * other electrical components. Design-in alters the thermal properties of SCD30
 * such that temperature and humidity offsets may occur when operating the
 * sensor in end-customer devices. Compensation of those effects is achievable
 * by writing the temperature offset found in continuous operation of the device
 * into the sensor.
 * The temperature offset value is saved in non-volatile memory. The last set
 * value will be used after repowering.
 *
 * @param temperature_offset    Temperature offset, unit [degrees Celsius * 100]
 *                              i.e. one tick corresponds to 0.01 degrees C
 *
 * @return                      0 if the command was successful, an error code
 *                              otherwise
 */
int16_t scd30_set_temperature_offset(uint16_t temperature_offset);

/**
 * scd30_set_altitude() - Set the altitude above sea level
 *
 * Measurements of CO2 concentration are influenced by altitude. When a value is
 * set, the altitude-effect is compensated. The altitude setting is disregarded
 * when an ambient pressure is set on the sensor with
 * scd30_start_periodic_measurement.
 * The altitude is saved in non-volatile memory. The last set value will be used
 * after repowering.
 *
 * @param altitude  altitude in meters above sea level, 0 meters is the default
 *                  value and disables altitude compensation
 *
 * @return          0 if the command was successful, an error code otherwise
 */
int16_t scd30_set_altitude(uint16_t altitude);

/**
 * scd30_get_automatic_self_calibration() - Read if the sensor's automatic self
 * calibration is enabled or disabled
 *
 * See scd30_enable_automatic_self_calibration() for more details.
 *
 * @param asc_enabled   Pointer to memory of where to set the self calibration
 *                      state. 1 if ASC is enabled, 0 if ASC disabled. Remains
 *                      untouched if return is non-zero.
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd30_get_automatic_self_calibration(uint8_t* asc_enabled);

/**
 * scd30_enable_automatic_self_calibration() - Enable or disable the sensor's
 * automatic self calibration
 *
 * When activated for the first time a period of minimum 7 days is needed so
 * that the algorithm can find its initial parameter set for ASC.
 * The sensor has to be exposed to fresh air for at least 1 hour every day.
 * Refer to the datasheet for further conditions
 *
 * ASC status is saved in non-volatile memory. When the sensor is powered down
 * while ASC is activated SCD30 will continue with automatic self-calibration
 * after repowering without sending the command.
 *
 * @param enable_asc    enable ASC if non-zero, disable otherwise
 *
 * @return              0 if the command was successful, an error code otherwise
 */
int16_t scd30_enable_automatic_self_calibration(uint8_t enable_asc);

/**
 * scd30_set_forced_recalibration() - Forcibly recalibrate the sensor to a known
 * value.
 *
 * Forced recalibration (FRC) is used to compensate for sensor drifts when a
 * reference value of the CO2 concentration in close proximity to the SCD30 is
 * available.
 *
 * For best results the sensor has to be run in a stable environment in
 * continuous mode at a measurement rate of 2s for at least two minutes before
 * applying the calibration command and sending the reference value.
 * Setting a reference CO2 concentration will overwrite the settings from ASC
 * (see scd30_enable_automatic_self_calibration) and vice-versa. The reference
 * CO2 concentration has to be in the range 400..2000 ppm.
 *
 * FRC value is saved in non-volatile memory, the last set FRC value will be
 * used for field-calibration after repowering.
 *
 * @param co2_ppm   recalibrate to this specific co2 concentration
 *
 * @return          0 if the command was successful, an error code otherwise
 */
int16_t scd30_set_forced_recalibration(uint16_t co2_ppm);

/**
 * Read out the serial number
 *
 * @param serial    the address for the result of the serial number.
 *                  --------------------------------------
 *                  THE BUFFER MUST HOLD AT LEAST 33 BYTES
 *                  --------------------------------------
 *                  Usage example:
 *                  ```
 *                  char scd30_serial[33];
 *                  if (scd30_read_serial(scd30_serial) == 0) {
 *                      printf("SCD30 serial: %s\n", scd30_serial);
 *                  } else {
 *                      printf("Error reading SCD30 serial\n");
 *                  }
 *                  ```
 *                  Contains a zero-terminated string.
 * @return          0 if the command was successful, else an error code.
 */
int16_t scd30_read_serial(char* serial);

#ifdef __cplusplus
}
#endif

#endif /* SCD30_H */


scd30/scd30.c
/*
 * Copyright (c) 2018, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "scd30.h"
#include "scd_git_version.h"
#include "sensirion_arch_config.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"

#ifdef SCD30_ADDRESS
static const uint8_t SCD30_I2C_ADDRESS = SCD30_ADDRESS;
#else
static const uint8_t SCD30_I2C_ADDRESS = 0x61;
#endif

#define SCD30_CMD_START_PERIODIC_MEASUREMENT 0x0010
#define SCD30_CMD_STOP_PERIODIC_MEASUREMENT 0x0104
#define SCD30_CMD_READ_MEASUREMENT 0x0300
#define SCD30_CMD_SET_MEASUREMENT_INTERVAL 0x4600
#define SCD30_CMD_GET_DATA_READY 0x0202
#define SCD30_CMD_SET_TEMPERATURE_OFFSET 0x5403
#define SCD30_CMD_SET_ALTITUDE 0x5102
#define SCD30_CMD_SET_FORCED_RECALIBRATION 0x5204
#define SCD30_CMD_AUTO_SELF_CALIBRATION 0x5306
#define SCD30_CMD_READ_SERIAL 0xD033
#define SCD30_SERIAL_NUM_WORDS 16
#define SCD30_WRITE_DELAY_US 20000

#define SCD30_MAX_BUFFER_WORDS 24
#define SCD30_CMD_SINGLE_WORD_BUF_LEN \
    (SENSIRION_COMMAND_SIZE + SENSIRION_WORD_SIZE + CRC8_LEN)

int16_t scd30_start_periodic_measurement(uint16_t ambient_pressure_mbar) {
    if (ambient_pressure_mbar &&
        (ambient_pressure_mbar < 700 || ambient_pressure_mbar > 1400)) {
        /* out of allowable range */
        return STATUS_FAIL;
    }

    return sensirion_i2c_write_cmd_with_args(
        SCD30_I2C_ADDRESS, SCD30_CMD_START_PERIODIC_MEASUREMENT,
        &ambient_pressure_mbar, SENSIRION_NUM_WORDS(ambient_pressure_mbar));
}

int16_t scd30_stop_periodic_measurement() {
    return sensirion_i2c_write_cmd(SCD30_I2C_ADDRESS,
                                   SCD30_CMD_STOP_PERIODIC_MEASUREMENT);
}

int16_t scd30_read_measurement(float* co2_ppm, float* temperature,
                               float* humidity) {
    int16_t ret;
    uint8_t data[3][4];

    ret =
        sensirion_i2c_write_cmd(SCD30_I2C_ADDRESS, SCD30_CMD_READ_MEASUREMENT);
    if (ret != STATUS_OK)
        return ret;

    ret = sensirion_i2c_read_words_as_bytes(SCD30_I2C_ADDRESS, &data[0][0],
                                            SENSIRION_NUM_WORDS(data));
    if (ret != STATUS_OK)
        return ret;

    *co2_ppm = sensirion_bytes_to_float(data[0]);
    *temperature = sensirion_bytes_to_float(data[1]);
    *humidity = sensirion_bytes_to_float(data[2]);

    return STATUS_OK;
}

int16_t scd30_set_measurement_interval(uint16_t interval_sec) {
    int16_t ret;

    if (interval_sec < 2 || interval_sec > 1800) {
        /* out of allowable range */
        return STATUS_FAIL;
    }

    ret = sensirion_i2c_write_cmd_with_args(
        SCD30_I2C_ADDRESS, SCD30_CMD_SET_MEASUREMENT_INTERVAL, &interval_sec,
        SENSIRION_NUM_WORDS(interval_sec));
    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);

    return ret;
}

int16_t scd30_get_data_ready(uint16_t* data_ready) {
    return sensirion_i2c_delayed_read_cmd(
        SCD30_I2C_ADDRESS, SCD30_CMD_GET_DATA_READY, 3000, data_ready,
        SENSIRION_NUM_WORDS(*data_ready));
}

int16_t scd30_set_temperature_offset(uint16_t temperature_offset) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD30_I2C_ADDRESS, SCD30_CMD_SET_TEMPERATURE_OFFSET,
        &temperature_offset, SENSIRION_NUM_WORDS(temperature_offset));
    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);

    return ret;
}

int16_t scd30_set_altitude(uint16_t altitude) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(SCD30_I2C_ADDRESS,
                                            SCD30_CMD_SET_ALTITUDE, &altitude,
                                            SENSIRION_NUM_WORDS(altitude));
    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);

    return ret;
}

int16_t scd30_get_automatic_self_calibration(uint8_t* asc_enabled) {
    uint16_t word;
    int16_t ret;

    ret = sensirion_i2c_read_cmd(SCD30_I2C_ADDRESS,
                                 SCD30_CMD_AUTO_SELF_CALIBRATION, &word,
                                 SENSIRION_NUM_WORDS(word));
    if (ret != STATUS_OK)
        return ret;

    *asc_enabled = (uint8_t)word;

    return STATUS_OK;
}

int16_t scd30_enable_automatic_self_calibration(uint8_t enable_asc) {
    int16_t ret;
    uint16_t asc = !!enable_asc;

    ret = sensirion_i2c_write_cmd_with_args(SCD30_I2C_ADDRESS,
                                            SCD30_CMD_AUTO_SELF_CALIBRATION,
                                            &asc, SENSIRION_NUM_WORDS(asc));
    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);

    return ret;
}

int16_t scd30_set_forced_recalibration(uint16_t co2_ppm) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd_with_args(
        SCD30_I2C_ADDRESS, SCD30_CMD_SET_FORCED_RECALIBRATION, &co2_ppm,
        SENSIRION_NUM_WORDS(co2_ppm));
    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);

    return ret;
}

int16_t scd30_read_serial(char* serial) {
    int16_t ret;

    ret = sensirion_i2c_write_cmd(SCD30_I2C_ADDRESS, SCD30_CMD_READ_SERIAL);
    if (ret)
        return ret;

    sensirion_sleep_usec(SCD30_WRITE_DELAY_US);
    ret = sensirion_i2c_read_words_as_bytes(SCD30_I2C_ADDRESS, (uint8_t*)serial,
                                            SCD30_SERIAL_NUM_WORDS);
    serial[2 * SCD30_SERIAL_NUM_WORDS] = '\0';
    return ret;
}

const char* scd30_get_driver_version() {
    return SCD_DRV_VERSION_STR;
}

uint8_t scd30_get_configured_address() {
    return SCD30_I2C_ADDRESS;
}

int16_t scd30_probe() {
    uint16_t data_ready;

    /* try to read data-ready state */
    return scd30_get_data_ready(&data_ready);
}
